### ISSUE TITLE
Support output as Markdown instead of html, for example, for github readme.md

### ISSUE THREAD

**Issue by MathewSachin** [dotnet/docfx#684]
DocFX does a great job in building static websites.
But, I want to know if there is a way to export the files in Markdown.
It would then be easy to put up on GitHub wikis or GitBook.

And by that I mean that I don't require features like search, Affix, Breadcrumb, navbar, footer, etc.

Here's what I tried:
- Renamed `*.html.primary.*` files to `*.md.primary.*`
- Modified partials and templates from html to markdown.
- Removed unnecessary features.

That was a good progress but with two major problems:
1. Markdown in conceptual documentation was getting converted to html. I really need to prevent this. Trying to add the *.md files to resources failed resulting in unresolved links.
2. Xml Documentation was being converted to html.

So, finally it seems that I want some way to convert that output html to markdown.

Any help is appreciated and BTW thank you for this awesome software.

**Comment by chenkennt** [comment]
@MathewSachin if you don't want markdown to be converted to html, what kind of functionality do you want from DocFX? You can directly send your markdown files to gitbook..

**Comment by MathewSachin** [comment]
@chenkennt I want DocFX to extract docs from xml documentation and augment my markdown files into it.

**Comment by vwxyzh** [comment]
@MathewSachin if you don't need merge overwrite document for managed reference, I think you can do it by following steps:

1. Create your [custom document processor](http://dotnet.github.io/docfx/tutorial/howto_build_your_own_type_of_documentation_with_custom_plug-in.html#create-a-document-processor):
    - Inherited from [ManagedReferenceDocumentProcessor](http://dotnet.github.io/docfx/api/Microsoft.DocAsCode.Build.ManagedReference.ManagedReferenceDocumentProcessor.html)
    - Override steps property different contract name (disable mark to html).
    - Give higher priority (disable default ManagedReferenceDocumentProcessor),
2. Create your custom template to create markdown file.

**Comment by MathewSachin** [comment]
Merging overwrite documents is necessary for me. But, still your idea seems implementable.

**Comment by MathewSachin** [comment]
Another problem arised that xrefs are not being resolved.

For example, see https://github.com/ManagedBass/ManagedBass.DocFX/tree/gitbook

**Comment by vwxyzh** [comment]
@MathewSachin Merging overwrite documents is based on html format, it is too hard to implement in markdown format.
And I don't have any idea for xref problem.

**Comment by vicancy** [comment]
@MathewSachin Your homepage looks pretty cool, would you willing to share it to http://dotnet.github.io/docfx/templates-and-plugins/templates-dashboard.html ?

**Comment by MathewSachin** [comment]
I would be happy to have it shared.

But, it is based on a previous version of DocFX.
Some recent changes in DocFX have resulted in some bugs in my website because my template is for the older versions.

The template really needs to be tweaked before sharing.
I will let you know once it is ready

**Comment by vicancy** [comment]
Different document type are sharing the same layout https://github.com/dotnet/docfx/tree/dev/src/docfx.website.themes/default/layout now, so I assume it is now much easier to change layout? 😄

**Comment by MathewSachin** [comment]
@vicancy I have created a new repository with the code for my DocFX template: https://github.com/MathewSachin/docfx-tmpl

You can add it to the Templates and Plugins page.
I would try to improve it further.

**Comment by Qevlarr** [comment]
I also request this feature. Documentation should be easy to find. Github (and other platforms) show a markdown readme right on a project's main page. It would be great if we can refer to generated documentation pages (again in markdown) with something like xref. The platform can style the markdown for me, that's not as important as having all the docs in one easy to find place.

**Comment by bloudraak** [comment]
I'm also in favor of Markdown... but for different reasons. We often have to incorporate our documentation with other content management systems that support markdown.  For example, we need to upload our APIs to Confluence, WordPress or some other system that supports Markdown.  The markdown has to be simple, limited to something like [CommonMark](https://commonmark.org), with simple code blocks, headings. Some of these CMS tools prefer the file structure to be in  a single directory (e.g. Confluence doesn't support subdirectories when linking between pages) and pages needs to be unique.  For API pages, the TOC on the left side is better served as its own page; each namespace and/or assembly may need its own page too. The class reference page can then link the namespace/assembly. 

It is ok to lose some features provided by the HTML output of DocFX, e.g. searching, styling etc, since that will all be taken care of by the CMS.

**Comment by superyyrrzz** [comment]
Hinted by #1253 , we can add a post step (Maybe [a PostProcessor](https://dotnet.github.io/docfx/tutorial/howto_add_a_customized_post_processor.html)) to convert HTML pages to Markdown files.

**Comment by open-collar** [comment]
This would be really useful as a means of posting generated documentation in Azure DevOps (as a "Code Wiki").

**Comment by CaptainCanonical** [comment]
I seem to be going in circles trying to find the answer to how to make docfx generate output in markdown instead of html. Is the answer what @vwxyzh posted on Sep 9, 2016 about creating your own custom processor,  etc? Is there no simpler solution yet?

**Comment by 0xff80ff** [comment]
It would be great if it will generate wiki pages instead of static site.

**Comment by lloydjatkinson** [comment]
Is there really no plan to generate markdown support?

**Comment by jsiegmund** [comment]
Came here in search for the same, also looking for a way to better integrate documentation we have in our Azure DevOps wiki along with documentation generated by docfx. Markdown output would be a perfect solution for us.

**Comment by daniellopezgarcia** [comment]
@jsiegmund I came here as well. After digging up a little, I found that .NET already generates structured (in xml format) documentation, which then can be parsed relatively easy into .md files. You just need to enable it from the build options and set the destination file. For my internal needs this may suffice for now but I'm sharing the script in case someone finds it useful. Apologies for the lack of interfacing but this is exactly what I need. Feel free to modify to suit your needs! (Change the input file or make it a input to the script)

### Read an input xml which follows [this format](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/), extract its classes and interfaces, and for each one, extract its associated methods and properties and print them. Concatenates multiline summaries into a single line, but tries to keep lists as they are intended.


```pwsh
# Using .xml files within this same directory, generates .md files containing MS Doc-like style documentation
# for classes and methods.

$documentationComments = [xml](Get-Content '.\Namespace.Project.xml')

# Obtain types and then its methods, constructors and constants.
$documentationComments.doc.members.ChildNodes | Where-Object { $_.name.StartsWith("T:")} | ForEach-Object {
    $element = $_
    $fullyQualifiedName = $element.name -replace "^T:",""
    $typeNamespace = $fullyQualifiedName.SubString(0, $fullyQualifiedName.LastIndexOf('.'))
    $typeName = $element.name.Split(".")[-1]
    Write-Host "Processing C# type $($typeName)"
    $mdFileName = "$($typeNamespace).$($typeName).md"
    $typeTrimmedSummary = $element.summary -replace "`n +","`n" -replace "^`n", "" -replace "`n$","" -replace "`n-","`n"

    # Add basic documentation (simple .xml to .md conversion)
    $AutogeneratedMDContent = @"

# $($typeName)

## Definition

*Namespace: $($typeNamespace)*

*Assembly: $($documentationComments.doc.assembly.name).dll*

$($typeTrimmedSummary)
"@

    # Add properties documentation
    if(($documentationComments.doc.members.ChildNodes | Where-Object { $_.name -match "^P:$($fullyQualifiedName -replace "\.","\.")\.[_a-zA-Z0-9]+"}).Count -gt 0){
        $AutogeneratedMDContent += "## Properties`n`n|Name|Description|`n|--|--|"
    }
    $documentationComments.doc.members.ChildNodes | Where-Object { $_.name -match "^P:$($fullyQualifiedName -replace "\.","\.")"} | ForEach-Object {        
        $propertyInfo = $_
        $propertyFullyQualifiedName = $propertyInfo.name -replace "^P:",""
        $propertyName = $propertyFullyQualifiedName.Split(".")[-1]
        Write-Host "Property $($propertyName) belongs to $($typeName)"
        $propertyTrimmedSummary = $propertyInfo.summary -replace "`n +","`n" -replace "^`n", "" -replace "`n$","" -replace "`n"," "
        $AutogeneratedMDContent += "`n|$($propertyName)|$($propertyTrimmedSummary)|"
    }

    # Add methods documentation
    if(($documentationComments.doc.members.ChildNodes | Where-Object { $_.name -match "^M:$($fullyQualifiedName -replace "\.","\.")\.[_a-zA-Z0-9]+\("}).Count -gt 0){
        $AutogeneratedMDContent += "## Methods`n`n|Name|Description|`n|--|--|"
    }
    $documentationComments.doc.members.ChildNodes | Where-Object { $_.name -match "^M:$($fullyQualifiedName -replace "\.","\.")\.[_a-zA-Z0-9]+\("} | ForEach-Object {
        $methodInfo = $_
        $methodFullyQualifiedName = $methodInfo.name -replace "^M:",""
        $methodSignature = $methodFullyQualifiedName -replace "^([0-9a-zA-Z]+\.)+","" -replace "System.String,","string, " -replace "System.String\)", "string)"
        Write-Host "Method $($_.name) belongs to $($typeName)"
        $MethodTrimmedSummary = $methodInfo.summary -replace "`n +","`n" -replace "^`n", "" -replace "`n$","" -replace "`n"," "
        $AutogeneratedMDContent += "`n|$($methodSignature)|$($MethodTrimmedSummary)|"
    }

    $AutogeneratedMDContent | Out-File $mdFileName
}
```

The good news is that extracting formatted comments into an structure is already done by .NET, not DocFX. The other good news is that converting from structured to simple .md is almost as straightforward as above (depends on what you need). Bad news is that its not done!

**Comment by yufeih** [comment]
https://github.com/ap0llo/mddocs seems like a promising tool to produce .NET API reference documentation in markdown format.

**Comment by daniellopezgarcia** [comment]
> https://github.com/ap0llo/mddocs seems like a promising tool to produce .NET API reference documentation in markdown format.

Seems very promising!
A little context is that we have docfx taking a wiki-like, git repository into producing nice html output. However, its a large, common wiki hosting multiple team's own style of documentation according to their needs, so it meets some flexibility for each team. However, its inputs are .md files, so that we do currently .cs -> .xml (by .NET) -> .md (by above script/promisind mddocs) -> .html (by docfx)

I'm going to check whether just adding the .xml into the wiki repo gets docfx to generate .html documentation.  The end goal is to have whatever format is required by our documentation process needs, I can think that for some people it can even mean PDF, but if .xml -> .html is directly possible (skipping .md altogether), then that also works just fine!

**Comment by Muchaszewski** [comment]
> https://github.com/ap0llo/mddocs seems like a promising tool to produce .NET API reference documentation in markdown format.

Unfortunately, mddocs will not work as it requires full to provide all assemblies and XML file generated. It also doesn't support tags like `see`, `inheritdoc` and `exceptions`. 

I would love to see fully supported MarkDown files generated from docfx as it has the most mature engine to take all the metadata.

**Comment by Ole-Visma** [comment]
More than 6 years since this "must have" feature without any solution...

**Comment by akovanev** [comment]
I've just created a small package which may slightly help with the issue (at least it helped for our project). I'll put it [here](https://github.com/akovanev/NetDocsProcessor) if you don't mind. 

However, it would be really nice to get a simple api retrieving the documentation data as a part of docfx.

**Comment by MPeli** [comment]
@yufeih, I see that you added this issue to [Working Set](https://github.com/dotnet/docfx/milestone/48). Does it mean that this feature request is currently being worked on? Thank you.

**Comment by yufeih** [comment]
Yes it is! I'll have some initial PR ready this week.

**Comment by MPeli** [comment]
@yufeih, thank you so much! We are really looking forward to using this new feature. Let me know if you need any help with testing.

**Comment by yufeih** [comment]
Experimental support of markdown output is added to `2.71.0` using the `outputFormat: markdown` config:

```
{
  "metadata": [
    {
      "outputFormat": "markdown",
    }
  ],
}
```
